packages/server/lib/StudioLifecycleManager.ts

```diff
@@ -0,0 +1,151 @@
+  private static readonly STUDIO_READY_EVENT = 'studio:ready'
+  private ctx: DataContext | null = null
+
+  /**
+   * Set the studio manager promise and register listeners
+   * @param studioManagerPromise Promise that resolves to a StudioManager
+   */
+  private setStudioManagerPromise (studioManagerPromise: Promise<StudioManager | null>): void {
+    this.studioManagerPromise = studioManagerPromise
+
+    // When the promise resolves, emit the studio:ready event with the studio manager
+    void studioManagerPromise.then((studioManager) => {
+      this.emit(StudioLifecycleManager.STUDIO_READY_EVENT, studioManager)
+    })
+  }
+
+  /**
+   * Initialize the studio manager and possibly set up protocol.
+   * Also registers this instance in the data context.
+   * @param projectId The project ID
+   * @param cloudDataSource The cloud data source
+   * @param cfg The project configuration
+   * @param debugData Debug data for the configuration
+   * @param ctx Data context to register this instance with
+   */
+  initializeStudioManager ({
+    projectId,
+    cloudDataSource,
+    cfg,
+    debugData,
+    ctx,
+  }: {
+    projectId?: string
+    cloudDataSource: CloudDataSource
+    cfg: Cfg
+    debugData: any
+    ctx: DataContext
+  }): void {
+    debug('Initializing studio manager')
+
+    const studioManagerPromise = getAndInitializeStudioManager({
+      projectId,
+      cloudDataSource,
+    }).then(async (studioManager) => {
+      if (studioManager.status === 'ENABLED') {
+        debug('Studio manager enabled, setting up protocol')
+        const protocolManager = new ProtocolManager()
+        const protocolUrl = routes.apiRoutes.captureProtocolCurrent()
+        const script = await api.getCaptureProtocolScript(protocolUrl)
+
+        await protocolManager.prepareProtocol(script, {
+          runId: 'studio',
+          projectId: cfg.projectId,
+          testingType: cfg.testingType,
+          cloudApi: {
+            url: routes.apiUrl,
+            retryWithBackoff: api.retryWithBackoff,
+            requestPromise: api.rp,
+          },
+          projectConfig: _.pick(cfg, ['devServerPublicPathRoute', 'port', 'proxyUrl', 'namespace']),
+          mountVersion: api.runnerCapabilities.protocolMountVersion,
+          debugData,
+          mode: 'studio',
+        })
+
+        studioManager.protocolManager = protocolManager
+        studioManager.isProtocolEnabled = true
+      }
+
+      return studioManager
+    }).catch((err) => {
+      debug('Error during studio manager setup: %o', err)
+
+      return null
+    })
+
+    this.setStudioManagerPromise(studioManagerPromise)
+
+    // Register this instance in the data context
+    ctx.update((data) => {
+      data.studioLifecycleManager = this
+    })
+  }
+
+  /**
+   * Check if studio manager is ready to be used
+   * @returns boolean indicating if studio manager promise exists
+   */
+  isStudioReady (): boolean {
+    return this.studioManagerPromise !== null
```

Comment:
What is the promise hasn't resolved, is studio still ready?

Reply:
not necessarily. I updated this to use a variable in the class to determine whether or not studio has been resolved instead